### PR TITLE
Introduce prettier

### PR DESCRIPTION
**Pull request by bluSCALE4**
Prettier is an opinionated code formatter. It makes it easy for people to contribute and helps bring consistent code aesthetics. Precommit hooks can be added to run prettier before commits. Prettier is one of those tools that many developers use and rely on to validate their code. If a project does not include it, causes large diffs when introducing minor code changes. Applying it universally avoids this issue and lowers the barrier of entry for developers to contribute.

[Prettier](https://prettier.io)

### CODE REVIEW

**Comment by klues** [comment]
Thanks for the PR! 👍 
As said in my mail, I like the idea of a general code formatter. However, I would like to keep the changes as minimal as possible, so I would like to configure Prettier to work similar to my current code formatter of IntelijIDEA. Here's my editorconfig file:
[editorconfig.zip](https://github.com/asterics/AsTeRICS-Grid/files/11037738/editorconfig.zip) (zipped because github doesn't allow hidden files)

This doc says that Prettier can import config from an .editorconfig file: https://prettier.io/docs/en/configuration.html#editorconfig
So I would suggest that you change the Prettier config in order to use the editorconfig file. For better maintainability it would be also nice (or even better) if this complex editorconfig file would be reduced to a simpler `.prettierrc` file, but until now I didn't find a way to do this automatically.

**Comment by bluSCALE4** [comment]
I reviewed the config and most of the properties are ignored. Prettier is pretty opinionated and does not have options for most customizations included. What Prettier strives for is consistency and maintaining a set line length. For example, it will happily inline an object or long strong or html element but once it grows larger than 80 characters, it will aggressively add newlines because it takes an all or nothing approach. On this point, there is no flexibility. 

Just to point one something that may have been missed, take a look at the commit diff and the side by side view on GitHub. With Prettier, the diff is extremely clear while the old code isn't. 

Some food for thought: Most web developers work in VSCode and some minorities use IntelliJ and Vim*. Real estate wise, the lowest common denominator is usually Vim where scrolling is not an option. They usually use 72 char width maxes though 80 char is more of the norm.

At the end of the day, consistent and predictable code is easier to digest and refactor. It makes code reviewing easier and caters to all IDEs vs most. 

Worst case, I can close this PR and focus only on the files I intend to refactor.

*Or one of its iterations

**Comment by klues** [comment]
>  it will happily inline an object or long string/html element but once it grows larger than 80 characters, it will aggressively add newlines because it takes an all or nothing approach. On this point, there is no flexibility.

OK, to my mind these newlines are a little bit too agressive. I don't think code blocks get more readable if they suddenly have the doubled amount of lines. And 80 chars seem to be really too little to my mind, I think something like 120 chars is appropriate. I don't think anybody will contribute with Vim (hello any future collaborator, you're welcome to prove me wrong). So I think it would be interesting how the Prettier result with [print width](https://prettier.io/docs/en/options.html#print-width) of 120 chars would be. I think the limitation to 80 chars was what bothered me most, for other things I think I would accustom myself to it.

What I also wasn't sure about is the git pre-commit hook. There are cases where I would like to be able to decide to not format specific lines according to Prettier, like this: https://github.com/asterics/AsTeRICS-Grid/blob/master/serviceWorker.js#L13 - it's something I'm generating and never change manually, therefore I don't need it to be split up to 25 lines. I'm prefering a formatter that helps me keep things formatted nicely, but doesn't force me to do things where I don't think it makes sense.

But yeah, until now I'm the only developer in this repo, and if other developers join I understand that it makes sense to have consistency, so I'm willing to rethink my opinions. I would give Prettier a try with 120 chars width, if you want to try it. I'm sorry if I've made you upset, which slightly appears so for me, since you've closed the PR deleted everthing you've did with a force push.

**Comment by klues** [comment]
> it will happily inline an object or long string/html element but once it grows larger than 80 characters, it will aggressively add newlines because it takes an all or nothing approach. On this point, there is no flexibility.

OK, to my mind these newlines are a little bit too agressive. I don't think code blocks get more readable if they suddenly have the doubled amount of lines. And 80 chars seem to be really too little to my mind, I think something like 120 chars is appropriate. I don't think anybody will contribute with Vim (hello any future collaborator, you're welcome to prove me wrong). So I think it would be interesting how the Prettier result with [print width](https://prettier.io/docs/en/options.html#print-width) of 120 chars would be. I think the limitation to 80 chars was what bothered me most, for other things I think I would accustom myself to it.

What I also wasn't sure about is the git pre-commit hook. There are cases where I would like to be able to decide to not format specific lines according to Prettier, like this: https://github.com/asterics/AsTeRICS-Grid/blob/master/serviceWorker.js#L13 - it's something I'm generating and never change manually, therefore I don't need it to be split up to 25 lines. I'm prefering a formatter that helps me keep things formatted nicely, but doesn't force me to do things where I don't think it makes sense.

But yeah, until now I'm the only developer in this repo, and if other developers join I understand that it makes sense to have consistency, so I'm willing to rethink my opinions. I would give Prettier a try with 120 chars width, if you want to try it. I'm sorry if I've made you upset, which slightly appears so for me, since you've closed the PR deleted everthing you've did with a force push.

**Comment by bluSCALE4** [comment]
Ah, no, my mistake. I just decided to work on something else and I was working on a fork master branch so I was forced to resync. I still have my local commits so I can cherry-pick into a dedicated branch. Silly of me to ask you to review the diffs only to resync a few hours later.

**Comment by klues** [comment]
Thanks, with 120 chars width I like it much better! 👍
I hope it doesn't break anything, but I'll trust the fact that prettier is widely used and known.

Some things:
* why the folder [vue-components](https://github.com/asterics/AsTeRICS-Grid/tree/master/src/vue-components) isn't formatted?
* I would prefer `"singleQuote": true` (but doubled are also ok, if you think it has advantages)
* regarding the pre-commit hook I think it's ok since I think `serviceWorker.js` is the only file where it doesn't make so much sense, however I think it also would be ok to format it.

**Comment by bluSCALE4** [comment]
I added serviceWorker to `.prettierignore` so it should be excluded. I have not yet applied prettier to vue files since when I did, it wasn't straight forward. There are two different formatters available and one looked terrible while the other looked great. The problem was that the pre-commit hook wasn't choosing that right one so I decided to hold off until it was decided to move ahead with Prettier or not.

File: webpack.config.js [hunk 1]
```diff
@@ -1,72 +1,76 @@
...
+            }
+        }
     };
 
     var resolve = {
         alias: {
             //objectmodel: "../../../node_modules/objectmodel/dist/object-model.js"
-            vue: 'vue/dist/vue.esm.js'
+            vue$: "vue/dist/vue.esm.js",
+            "@": path.resolve("src")
```
Comment (bluSCALE4) [inline webpack.config.js:52]: This makes it so you can reference files easier. `import {constants} from "../util/constants";` -> `import {constants} from "@/js/util/constants";`

Though I believe I wrote it wrong. It should look like this:
```
            "@": path.resolve(__dirname, 'src'),
            "@assets": path.resolve(__dirname, 'app')
```

**Comment by bluSCALE4** [comment]
https://www.browserstack.com/contact?ref=footer#open-source￼
Contact us
browserstack.com

Looks like you can request free access right from their site.

File: webpack.config.js (hunk 1 above)
Comment (klues) [reply webpack.config.js:52]: ok, but please keep this PR focussed on Prettier and let's don't add or change other things.

**Comment by klues** [comment]
> There are two different formatters available and one looked terrible while the other looked great. The problem was that the pre-commit hook wasn't choosing that right one so I decided to hold off until it was decided to move ahead with Prettier or not.

OK, but how do you select different formatters within the IDE? I thought Prettier is all about consistency and there's no freedom of choosing different formatters of your choice?!

**Comment by klues** [comment]
I've added your Prettier config to the master branch, updated my IDE to use it, formatted the `src/js` folder with it and checked the diff to your branch locally. It was very similar, but at some points there where differences, I don't know why. I've also used `singleQuote: true`.

Please merge the current master to your master and try to reformat `src/js` to see if the results are the same. If yes, you could also format `src/css` and `src/vue-components`, I'll check if it results in the same format in my IDE and if yes, we can merge this PR.

**Pull request closed without merging**